File: blueluak-core/src/main/java/net/blueva/luak/lib/ResourceFinder.java
/******************************************************************************
 *  ____  _            _                    _
 * | __ )| |_   _  ___| |   _   _  __ _    | |
 * |  _ \| | | | |/ _ \ |  | | | |/ _` |_  | |
 * | |_) | | |_| |  __/ |__| |_| | (_| | |_| |
 * |____/|_|\__,_|\___|_____\__,_|\__,_|\___/
 *
 *  BlueLuaK
 *  https://github.com/BluevaDevelopment/BlueLuaK/t
 *
 *  Based on LuaJ (https://luaj.org)
 *  Original work Copyright (c) 2009 Luaj.org
 *  Modifications Copyright (c) 2026 Blueva Development
 *
 *  SPDX-License-Identifier: MIT
 ******************************************************************************/
package net.blueva.luak.lib;

import java.io.InputStream;

import net.blueva.luak.Globals;

/** 
 * Interface for opening application resource files such as scripts sources.  
 * <p>
 * This is used by required to load files that are part of 
 * the application, and implemented by BaseLib
 * for both the Jme and Jse platforms. 
 * <p>
 * The Jme version of base lib {@link BaseLib} 
 * implements {@link Globals#finder} via {@link Class#getResourceAsStream(String)}, 
 * while the Jse version {@link net.blueva.luak.lib.jse.JseBaseLib} implements it using {@link java.io.File#File(String)}.
 * <p>
 * The io library does not use this API for file manipulation.
 * <p>
 * @see BaseLib
 * @see Globals#finder
 * @see net.blueva.luak.lib.jse.JseBaseLib
 * @see net.blueva.luak.lib.jme.JmePlatform
 * @see net.blueva.luak.lib.jse.JsePlatform 
 */
public interface ResourceFinder {
	
	/** 
	 * Try to open a file, or return null if not found.
	 * 
	 * @see net.blueva.luak.lib.BaseLib
	 * @see net.blueva.luak.lib.jse.JseBaseLib
	 * 
	 * @param filename
	 * @return InputStream, or null if not found. 
	 */
	public InputStream findResource( String filename );
}

File: blueluak-core/src/commonMain/kotlin/net/blueva/luak/lib/ResourceFinder.kt
/******************************************************************************
 *  ____  _            _                _  __
 * | __ )| |_   _  ___| |   _   _  __ _| |/ /
 * |  _ \| | | | |/ _ \ |  | | | |/ _` | ' /
 * | |_) | | |_| |  __/ |__| |_| | (_| | . \
 * |____/|_|\__,_|\___|_____\__,_|\__,_|_|\_\
 *
 *  BlueLuaK
 *  https://github.com/BluevaDevelopment/BlueLuaK
 *
 *  Based on LuaJ (https://luaj.org)
 *  Original work Copyright (c) 2009 Luaj.org
 *  Modifications Copyright (c) 2026 Blueva Development
 *
 *  SPDX-License-Identifier: MIT
 ******************************************************************************/
package net.blueva.luak.lib

import net.blueva.luak.io.InputStream

/**
 * Interface for opening application resource files such as scripts sources.
 * 
 * 
 * This is used by required to load files that are part of
 * the application, and implemented by BaseLib
 * for both the Jme and Jvm platforms.
 * 
 * 
 * The Jme version of base lib [BaseLib]
 * implements [Globals.finder] via [Class.getResourceAsStream],
 * while the Jvm version [net.blueva.luak.lib.jvm.JvmBaseLib] implements it using [java.io.File.File].
 * 
 * 
 * The io library does not use this API for file manipulation.
 * 
 * 
 * @see BaseLib
 * 
 * @see Globals.finder
 * 
 * @see net.blueva.luak.lib.jvm.JvmBaseLib
 * 
 * @see net.blueva.luak.lib.jme.JmePlatform
 * 
 * @see net.blueva.luak.lib.jvm.JvmPlatform
 */
interface ResourceFinder {
    /**
     * Try to open a file, or return null if not found.
     * 
     * @see net.blueva.luak.lib.BaseLib
     * 
     * @see net.blueva.luak.lib.jvm.JvmBaseLib
     * 
     * 
     * @param filename
     * @return InputStream, or null if not found.
     */
    fun findResource(filename: String?): InputStream?
}